### PR TITLE
dev/sg: lint against os.Exit and signal.Notify in sg

### PR DESCRIPTION
**Pull request by bobheadxi**
Follow-up to https://github.com/sourcegraph/sourcegraph/pull/35727

## Test plan

image

### CODE REVIEW

File: dev/sg/internal/repo/repo.go [hunk 1]
```diff
@@ -39,10 +39,6 @@ type DiffHunk struct {
 }
 
 func (s *State) GetDiff(paths string) (Diff, error) {
-	if paths == "" {
-		paths = "**/*"
```
Comment (bobheadxi) [inline dev/sg/internal/repo/repo.go:43]: I realized this isn't needed

Comment (bobheadxi) [reply dev/sg/internal/repo/repo.go]: spun off into https://github.com/sourcegraph/sourcegraph/pull/35761

File: dev/sg/linters/go_checks.go [hunk 2]
```diff
@@ -1,9 +1,42 @@
...
 	goFmt          = lint.RunScript("Go format", "dev/check/gofmt.sh")
 	goLint         = lint.RunScript("Go lint", "dev/check/go-lint.sh")
 	goDBConnImport = lint.RunScript("Go pkg/database/dbconn", "dev/check/go-dbconn-import.sh")
 )
+
+func lintSGExit() lint.Runner {
+	const header = "Lint dev/sg exit signals"
+
+	return func(ctx context.Context, s *repo.State) *lint.Report {
+		diff, err := s.GetDiff("dev/sg/***.go")
```
Comment (mrnugget) [inline dev/sg/linters/go_checks.go:22]: Shouldn't it be this? Might be wrong though, maybe this is the new way of doing "give me all the things"-globs 😄 

```suggestion
		diff, err := s.GetDiff("dev/sg/**/*.go")
```

File: dev/sg/linters/go_checks.go [hunk 3]
```diff
@@ -1,9 +1,42 @@
...
+		if err != nil {
+			return &lint.Report{Header: header, Err: err}
+		}
+
+		mErr := diff.IterateHunks(func(file string, hunk repo.DiffHunk) error {
+			if strings.HasPrefix(file, "dev/sg/interrupt") || strings.HasPrefix(file, "dev/sg/linters/go_checks.go") {
```
Comment (mrnugget) [inline dev/sg/linters/go_checks.go:28]:
```suggestion
			if strings.HasPrefix(file, "dev/sg/interrupt") || file == "dev/sg/linters/go_checks.go" {
```

**Review by mrnugget — APPROVED** [review]

**Comment by mrnugget** [comment]
One thought I want to add (and this is not a blocker): let's be mindful about adding linters and checks. There's a very real possibility that some things a linter prevents will never happen again and were one-off problems. By adding a linter, though, we increase CI/lint runtime ever so slightly and increase a chance of false positives.

Example: if someone writes in a Go comment `// we shouldn't use os.Exit here` this linter would fail the PR, the author would have to rewrite their comment and re-push and wait for CI again _or_ put more effort into the check here to make sure it ignores comments. Compare that to the chance of someone using `os.Exit` in `dev/sg` again and it slipping through code review.

Again, not blocking if you feel strongly about this, but I'm wondering whether we should have a "it has to have happened 5 times for a linter to be introduced" rule :)

File: dev/sg/linters/go_checks.go (hunk 2 above)
Comment (bobheadxi) [reply dev/sg/linters/go_checks.go:22]: Crazy, but I actually wasn't able to get `**/*.go` to give me everything - `dev/sg/main.go` would refuse to show up 😬

**Pull request merged**